webhook/resources/volume/mutator.go

```diff
@@ -35,14 +43,149 @@ func (v *volumeMutator) Resource() admission.Resource {
+	}
+
+	size := volume.Spec.Size
+	if volume.Spec.FromBackup != "" {
+		bName, bvName, _, err := backupstore.DecodeBackupURL(volume.Spec.FromBackup)
+		if err != nil {
+			return nil, werror.NewInvalidError(fmt.Sprintf("cannot get backup and volume name from backup URL %v: %v", volume.Spec.FromBackup, err), "")
+		}
+
+		bv, err := v.ds.GetBackupVolumeRO(bvName)
+		if err != nil {
+			return nil, werror.NewInvalidError(fmt.Sprintf("cannot get backup volume %s: %v", bvName, err), "")
+		}
+		if bv != nil && bv.Status.BackingImageName != "" {
+			if volume.Spec.BackingImage == "" {
+				volume.Spec.BackingImage = bv.Status.BackingImageName
+				patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/spec/backingImage", "value": "%s"}`, bv.Status.BackingImageName))
+				logrus.Debugf("Since the backing image is not specified during the restore, "+
+					"the previous backing image %v used by backup volume %v will be set for volume %v creation",
+					bv.Status.BackingImageName, bvName, name)
+			}
+			bi, err := v.ds.GetBackingImage(volume.Spec.BackingImage)
+			if err != nil {
+				return nil, werror.NewInvalidError(err.Error(), "")
+			}
+			// Validate the checksum only when the chosen backing image name is the same as the record in the backup volume.
+			// If user picks up a backing image different from `backupVolume.BackingImageName`, there is no need to do verification.
+			if volume.Spec.BackingImage == bv.Status.BackingImageName {
+				if bv.Status.BackingImageChecksum != "" && bi.Status.Checksum != "" &&
+					bv.Status.BackingImageChecksum != bi.Status.Checksum {
+					return nil, werror.NewInvalidError(fmt.Sprintf("backing image %v current checksum doesn't match the recoreded checksum in backup volume", volume.Spec.BackingImage), "")
+				}
+			}
+		}
+
+		backup, err := v.ds.GetBackupRO(bName)
+		if err != nil {
+			return nil, werror.NewInvalidError(fmt.Sprintf("cannot get backup %s: %v", bName, err), "")
+		}
+
+		logrus.Infof("Override size of volume %v to %v because it's from backup", name, backup.Status.VolumeSize)
+		// formalize the final size to the unit in bytes
+		size, err = util.ConvertSize(backup.Status.VolumeSize)
+		if err != nil {
+			return nil, werror.NewInvalidError(fmt.Sprintf("get invalid size for volume %v: %v", backup.Status.VolumeSize, err), "")
+		}
+	}
+
+	size = util.RoundUpSize(size)
+	patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/spec/size", "value": "%v"}`, strconv.FormatInt(size, 10)))
+
+	defaultEngineImage, _ := v.ds.GetSettingValueExisted(types.SettingNameDefaultEngineImage)
+	if defaultEngineImage == "" {
+		return nil, werror.NewInvalidError("BUG: Invalid empty Setting.EngineImage", "")
+	}
+	patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/spec/engineImage", "value": "%s"}`, defaultEngineImage))
+
+	// Add backup volume name label to the restore/DR volume
+	if volume.Spec.FromBackup != "" {
```

Comment:
Can we move this part as part of the same condition above? `volume.Spec.FromBackup != ""`